iceoryx_utils/include/iceoryx_utils/internal/relocatable_pointer/base_relative_ptr.hpp

```diff
@@ -0,0 +1,184 @@
+/// @note It should be noted that relocating a memory segment will invalidate relative pointers, i.e. relative pointers
+/// are NOT relocatable. This is because the registration mechanism cannot be automatically informed about the copy of a
+/// whole segment, such a segment would have to be registered on its own (and the original segment deregistered).
+class BaseRelativePointer
+{
+  public:
+    using id_t = uint64_t;
+    using ptr_t = void*;
+    using const_ptr_t = const void* const;
+    using offset_t = std::uintptr_t;
+
+    /// @brief constructs a BaseRelativePointer pointing to ptr in a segment identified by id
+    /// @param[in] ptr is the pointee
+    /// @param[in] id is the unique id of the segment
+    BaseRelativePointer(ptr_t ptr, id_t id) noexcept;
+
+    /// @brief constructs a BaseRelativePointer from a given offset and segment id
+    /// @param[in] offset is the offset
+    /// @param[in] id is the unique id of the segment
+    BaseRelativePointer(offset_t offset, id_t id) noexcept;
+
+    /// @brief constructs a BaseRelativePointer pointing to ptr
+    /// @param[in] ptr is the pointee
+    BaseRelativePointer(ptr_t ptr = nullptr) noexcept;
+
+    /// @brief copy constructor
+    /// @param[in] other is the copy origin
+    BaseRelativePointer(const BaseRelativePointer& other) noexcept;
+
+    /// @brief move constructor
+    /// @param[in] other is the move origin
+    BaseRelativePointer(BaseRelativePointer&& other) noexcept;
+
+    /// @brief copy assignment
+    /// @param[in] other is the copy origin
+    /// @return a reference to self
+    BaseRelativePointer& operator=(const BaseRelativePointer& other) noexcept;
+
+    /// @brief assigns the BaseRelativePointer to point to ptr
+    /// @param[in] ptr is the pointee
+    /// @return reference to self
+    BaseRelativePointer& operator=(void* ptr) noexcept;
+
+    /// @brief move assignment
+    /// @param[in] other is the move origin
+    /// @return a reference to self
+    BaseRelativePointer& operator=(BaseRelativePointer&& other) noexcept;
+
+    /// @brief access to the underlying object
+    /// @return a poiter to the underlying object
+    ptr_t get() const noexcept;
+
+    /// @brief returns the id which identifies the segment
+    /// @return the id which identifies the segment
+    id_t getId() const noexcept;
+
+    /// @brief returns the offset
+    /// @return the offset
+    offset_t getOffset() const noexcept;
+
+    /// @brief get the base pointer associated with this' id
+    /// @return the registered base pointer
+    ptr_t getBasePtr() const noexcept;
+
+    /// @brief registers a memory segment at ptr with size of a new id
+    /// @param[in] ptr starting address of the segment to be registered
+    /// @param[in] size is the size of the segment
+    /// @return id it was registered to
+    static id_t registerPtr(const ptr_t ptr, uint64_t size = 0U) noexcept;
+
+    /// @brief registers a memory segment at ptr with size of given id
```

Comment:
tries to register a memory segment with a given size starting at ptr to a given id